spicy.go

```diff
@@ -0,0 +1,171 @@
+func FormatProof(idx int64, p tlog.RecordProof, signedCheckpoint []byte) []byte {
+	return formatProof(idx, p, signedCheckpoint, nil, false)
+}
+
+// FormatProofWithHint formats a tlog record inclusion proof (a "spicy signature")
+// for the record at index idx with proof p and signed checkpoint signedCheckpoint,
+// including a hint.
+//
+// The returned byte slice is encoded according to c2sp.org/tlog-proof.
+func FormatProofWithHint(idx int64, hint []byte, p tlog.RecordProof, signedCheckpoint []byte) []byte {
+	return formatProof(idx, p, signedCheckpoint, hint, true)
+}
+
+func formatProof(idx int64, p tlog.RecordProof, signedCheckpoint []byte, hint []byte, withHint bool) []byte {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "c2sp.org/tlog-proof\n")
+	fmt.Fprintf(&buf, "index %d\n", idx)
+	if withHint {
+		fmt.Fprintf(&buf, "hint %s\n", base64.StdEncoding.EncodeToString([]byte(hint)))
+	}
+	for _, h := range p {
+		fmt.Fprintf(&buf, "%s\n", h)
+	}
+	fmt.Fprintf(&buf, "\n")
+	buf.Write(signedCheckpoint)
+	return buf.Bytes()
+}
+
+// VerifyRecordError is returned by [VerifyProof] when the inclusion proof does
+// not verify. It can be used to diagnose the issue or print a better error
+// message. All of its fields are unauthenticated and must not be trusted.
+type VerifyRecordError struct {
+	Index int64
+	Hint  []byte
+}
+
+func (e *VerifyRecordError) Error() string {
+	return fmt.Sprintf("tlog record inclusion proof verification failed for index %d", e.Index)
+}
+
+// VerifyProof verifies a proof (a "spicy signature" encoded according to
+// c2sp.org/tlog-proof) for a record hash rh (generally produced with
+// [tlog.RecordHash]).
+//
+// The origin must match the log's origin, and the open function will be used to
+// verify the signed checkpoint included in the proof. If open returns an error,
+// it is returned directly. If the proof is valid but does not verify the record
+// hash rh at the given index, a *[VerifyRecordError] is returned.
+func VerifyProof(origin string, open func([]byte) (*note.Note, error), rh tlog.Hash, proof []byte) error {
+	hdr, rest, ok := strings.Cut(string(proof), "\n")
+	if !ok || hdr != "c2sp.org/tlog-proof" {
+		return errors.New("malformed tlog proof: missing header, this may not be a tlog proof")
+	}
+	rest, ok = strings.CutPrefix(rest, "index ")
+	if !ok {
+		return errors.New("malformed tlog proof")
+	}
+	s, rest, ok := strings.Cut(rest, "\n")
+	if !ok {
+		return errors.New("malformed tlog proof")
+	}
+	idx, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return fmt.Errorf("malformed tlog proof: invalid index: %w", err)
+	}
+	if idx < 0 {
+		return fmt.Errorf("malformed tlog proof: negative index")
+	}
+	var hint []byte
+	if strings.HasPrefix(rest, "hint ") {
+		rest, ok = strings.CutPrefix(rest, "hint ")
+		if !ok {
+			return errors.New("malformed tlog proof")
+		}
+		s, rest, ok = strings.Cut(rest, "\n")
+		if !ok {
+			return errors.New("malformed tlog proof")
```

Comment:
most instances of these that don't already have a nice comment (and good on all those!) seem like they could probably be `return errors.New("malformed tlog proof: unexpected early termination (during {{part}})")`